/*
    Una Company (azienda) è legata ad un unico Business Admin che risulta abilitato ad inserire le informazioni riguardo la sua azienda.
    In particolare, di un'azienda conosciamo il codice fiscale aziendale, il nome, informazioni utili come locazione, area, indirizzo, ecc. 
    La lista dei prodotti che produce, i suoi ruoli nella supply chain. Se l'azienda ha come ruolo Production, è possibile specificare le informazioni
    legate ai suoi campi di produzione (Fields).
*/

syntax = "proto3";

message Company {

    // Documento, contiene un riferimento (url) allo storage off-chain e l'hash per l'integrità
    message Document {
        string url = 1;

        string hash = 2;

        string description = 3;
    }
    
    // Informazioni sul campo
    message Field {
        // Id univoco
        string id_field = 1;
        
        uint32 area = 2;

        repeated Document docs = 3;
    }

    // Locazione
    message Location {
        string latitude = 1;
        string longitude = 2;
    }

    // Informazioni utili
    message Info {
        Location location = 1;

        string area = 2;

        // Breve descrizione
        string description = 3;
    }

    // Lista operatori aziendali
    message Reporter {

        // Chiave pubblica operatore
        string op_pk = 1;

        // id evento
        string id = 2;

        uint64 timestamp = 3;
    }

    // Ruolo dell'azienda nella supply chain
    enum Role {
        // maybe aggiungere input? per tracciare materie prime
        PRODUCTION = 0;
        SUPPLIER = 1;
        PROCESSING = 2;
        DISTRIBUTION = 3;
        RETAILING = 4;
    }

    // Codice Fiscale aziendale (univoco)
    string company_fc = 1;

    // Nome
    string name = 2;

    // Informazioni utili
    Info info = 3;

    // Lista dei ruoli dell'azienda all'interno della supply chain. 
    // In base al ruolo sono definiti gli eventi che un Operator è abilitato a registrare nell'azienda
    // I prodotti che l'azienda può produrre, ecc.
    // Es. un azienda può essere sia Production che Supplier
    repeated Role roles = 4;

    // Chiave pubblica dell'amministratore
    string adm_pk = 5;

    // Lista degli operatori dell'azienda
    repeated Reporter operators = 6;

    // Lista campi (vale solo se Role = Production)
    repeated Field fields = 7;
}

message CompanyContainer {
    repeated Company entries = 1;
}